Mlem/Logic/InstanceMetadataParser.swift

```diff
@@ -0,0 +1,142 @@
+    enum ParsingError: Error {
+        case invalidData
+        case requiredHeaderMissing
+        case requestedIndexNotPresent
+        case noInstancesFound
+    }
+    
+    private enum Field: String, CaseIterable {
+        case instance = "Instance"
+        case newCommunities = "NC"
+        case newUsers = "NU"
+        case federated = "Fed"
+        case adultContent = "Adult"
+        case downvotes = "↓V"
+        case users = "Users"
+        case blockingInstanceCount = "BI"
+        case blockedByCount = "BB"
+        case uptime = "UT"
+        case version = "Version"
+    }
+    
+    private struct IndexContainer {
+        private var dictionary = [Field: Int]()
+        
+        var count: Int { dictionary.count }
+        
+        mutating func add(_ field: Field, for index: Int) {
+            dictionary[field] = index
+        }
+        
+        func index(of field: Field) throws -> Int {
+            guard let index = dictionary[field] else {
+                throw ParsingError.requestedIndexNotPresent
+            }
+            
+            return index
+        }
+    }
+    
+    // MARK: - Initialisation
+    
+    private init() { /* This struct is not designed to be constructed, use it's static method `.parse(from: ...)` */ }
+    
+    // MARK: - Public Methods
+    
+    static func parse(from data: Data) throws -> [InstanceMetadata] {
+        guard let string = String(data: data, encoding: .utf8), !string.isEmpty else {
+            throw ParsingError.invalidData
+        }
+        
+        var lines = string.split(separator: "\n")
+        let headerFields = lines.removeFirst().split(separator: ",").map { String($0) }
+        let indexes = createIndexContainer(from: headerFields)
+        
+        guard indexes.count == Field.allCases.count else {
+            throw ParsingError.requiredHeaderMissing
+        }
+        
+        let metadata = lines.compactMap { try? parse($0, using: indexes) }
+        
+        guard !metadata.isEmpty else {
+            throw ParsingError.noInstancesFound
+        }
+        
+        return metadata
+    }
+    
+    // MARK: - Private Methods
+    
+    private static func createIndexContainer(from headerFields: [String]) -> IndexContainer {
+        headerFields.reduce(into: IndexContainer()) { container, identifier in
+            guard
+                let field = Field(rawValue: identifier),
+                let index = index(of: field, in: headerFields)
+            else {
+                return
+            }
+            
+            container.add(field, for: index)
+        }
+    }
+    
+    private static func index(of field: Field, in fields: [String]) -> Int? {
+        guard let index = fields.firstIndex(of: field.rawValue) else {
+            return nil
+        }
+        
+        return Int(index)
+    }
+    
+    private static func parse(_ line: Substring, using indexes: IndexContainer) throws -> InstanceMetadata? {
```

Comment:
*nit*

Rename to `parseLine` so we don't have two methods named `parse` that do different things